posthog/temporal/data_modeling/workflows/execute_dag.py

```diff
@@ -0,0 +1,337 @@
+    return edge_lookup
+
+
+def _get_dependent_lookup(edge_lookup: dict):
+    # builds the inverse of the edge_lookup in the DAG object
+    dependents = defaultdict(set)
+    for target, sources in edge_lookup.items():
+        for source in sources:
+            dependents[source].add(target)
+    return dependents
+
+
+def _get_downstream_lookup(edge_lookup: dict):
+    downstreams = _get_dependent_lookup(edge_lookup)
+
+    def _get_all_downstream(node_id: str, visited: set[str]) -> set[str]:
+        if node_id in visited:
+            return set()
+        visited.add(node_id)
+        result = set(downstreams[node_id])
+        for downstream in list(downstreams[node_id]):
+            result.update(_get_all_downstream(downstream, visited))
+        return result
+
+    for node in list(downstreams.keys()):
+        visited: set[str] = set()
+        downstreams[node] = _get_all_downstream(node, visited)
+
+    return downstreams
+
+
+def _dag_execution_levels(
+    team_id: int,
+    dag_id: str,
+    nodes: list[str],
+    edge_lookup: dict,
+) -> list[list[str]]:
+    """Compute execution levels using kahn's topological sort."""
+    # Initialize in_degree for all nodes, defaulting to 0 for nodes with no dependencies
+    in_degree = {node_id: len(edge_lookup.get(node_id, set())) for node_id in nodes}
+    # inverse of the edge_lookup
+    dependents = _get_dependent_lookup(edge_lookup)
+    levels: list[list[str]] = []
+    remaining = nodes.copy()
+    while remaining:
+        current_level = [node_id for node_id in remaining if in_degree[node_id] == 0]
+        if not current_level:
+            # the only cases where this is possible are an empty DAG or a cycle in the DAG
+            raise EmptyDAGOrCycleError(f"DAG is either empty or contains a cycle: team={team_id} dag={dag_id}")
+        levels.append(current_level)
+        for node_id in current_level:
+            remaining.remove(node_id)
+            for dependent in dependents[node_id]:
+                if dependent in remaining:
+                    in_degree[dependent] -= 1
+    return levels
+
+
+@temporalio.workflow.defn(name="execute-dag")
+class ExecuteDAGWorkflow(PostHogWorkflow):
+    """Temporal workflow to orchestrate materialization of all nodes in a DAG.
+
+    This workflow:
+    1. Fetches the DAG structure (nodes and edges)
+    2. Computes execution levels based on dependencies
+    3. Executes materialization workflows in parallel for each level
+    4. Tracks success/failure and skips downstream nodes on failure
+    """
+
+    @staticmethod
+    def parse_inputs(inputs: list[str]) -> ExecuteDAGInputs:
+        """Parse inputs from the management command CLI."""
+        loaded = json.loads(inputs[0])
+        return ExecuteDAGInputs(**loaded)
+
+    @temporalio.workflow.run
+    async def run(self, inputs: ExecuteDAGInputs) -> ExecuteDAGResult:
+        temporalio.workflow.logger.info("Starting DAGOrchestratorWorkflow", extra=inputs.properties_to_log)
```

Comment:
Do these logs get output via our usual logger? I've never actually used the inbuilt temporal logger before